### PR TITLE
feat(Liquidation: Undercollateral): finish Undercollateral script and test

### PR DESCRIPTION
**Pull request by cosinlink**

### CODE REVIEW

File: contracts/toCKB-typescript/src/switch/liquidation_undercollateral/mod.rs [hunk 1]
```diff
@@ -1,7 +1,85 @@
...
+            btc_lot_size.get_sudt_amount()
+        }
+        XChainKind::Eth => {
+            let eth_lot_size = input_data.get_eth_lot_size()?;
+            eth_lot_size.get_sudt_amount()
+        }
+    };
+
+    // get X/CKB price from witness
+    let witness_args = load_witness_args(0, Source::Input)?.input_type();
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_undercollateral/mod.rs:71]:
```suggestion
    let witness_args = load_witness_args(0, Source::GroupInput)?.input_type();
```

File: contracts/toCKB-typescript/src/switch/liquidation_undercollateral/mod.rs [hunk 2]
```diff
@@ -1,7 +1,85 @@
...
+    if witness_args.is_none() {
+        return Err(Error::WitnessInvalid);
+    }
+    let witness_bytes: Bytes = witness_args.to_opt().unwrap().unpack();
+    let price: u8 = witness_bytes[0];
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_undercollateral/mod.rs:76]: use `u128`

File: contracts/toCKB-typescript/src/switch/liquidation_undercollateral/mod.rs [hunk 3]
```diff
@@ -1,7 +1,85 @@
...
+
+    // LIQUIDATION_COLLATERAL_PERCENT means min liquidation threshold of collateral/lot_amount
+    if singer_collateral * (price as u128) >= lot_amount * (LIQUIDATION_COLLATERAL_PERCENT as u128)
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_undercollateral/mod.rs:79]:
```suggestion
    if singer_collateral * (price as u128) * 100 >= lot_amount * (LIQUIDATION_COLLATERAL_PERCENT as u128)
```

File: tests/src/toCKB_typescript/tests/liquidation_undercollateral.rs [hunk 4]
```diff
@@ -1 +1,188 @@
...
+    );
+
+    let cycles = context
+        .verify_tx(&tx, MAX_CYCLES)
+        .expect("pass verification");
+    println!("consume cycles: {}", cycles);
+}
+
+#[test]
+fn test_wrong_price() {
```
Comment (mkxbl) [inline tests/src/toCKB_typescript/tests/liquidation_undercollateral.rs:43]:
```suggestion
fn test_wrong_undercollateral() {
```

File: contracts/toCKB-typescript/src/switch/liquidation_undercollateral/mod.rs (hunk 1 above)
Comment (cosinlink) [reply contracts/toCKB-typescript/src/switch/liquidation_undercollateral/mod.rs]: fixed

File: contracts/toCKB-typescript/src/switch/liquidation_undercollateral/mod.rs (hunk 2 above)
Comment (cosinlink) [reply contracts/toCKB-typescript/src/switch/liquidation_undercollateral/mod.rs]: fixed

File: contracts/toCKB-typescript/src/switch/liquidation_undercollateral/mod.rs (hunk 3 above)
Comment (cosinlink) [reply contracts/toCKB-typescript/src/switch/liquidation_undercollateral/mod.rs]: fixed

File: tests/src/toCKB_typescript/tests/liquidation_undercollateral.rs [hunk 5]
```diff
@@ -1 +1,189 @@
...
+        .type_(Some(toCKB_typescript.clone()).pack())
+        .lock(always_success_lockscript)
+        .build()];
+    let outputs_data = vec![output_toCKB_data; 1];
+
+    // witness
+    let data: [u8; 16] = price.to_le_bytes();
+    let witness = WitnessArgs::new_builder()
+        .input_type(Some(Bytes::copy_from_slice(data.as_ref())).pack())
+        .build(); // build transaction
```
[the remark this answers is not in the document]
Comment (cosinlink) [reply tests/src/toCKB_typescript/tests/liquidation_undercollateral.rs]: fixed

File: tests/src/toCKB_typescript/tests/liquidation_undercollateral.rs (hunk 4 above)
Comment (cosinlink) [reply tests/src/toCKB_typescript/tests/liquidation_undercollateral.rs]: fixed

File: contracts/toCKB-typescript/src/switch/liquidation_undercollateral/mod.rs [hunk 6]
```diff
@@ -1,7 +1,84 @@
...
+        .as_ref()
+        .expect("inputs should contain toCKB cell");
+    let output_data = toCKB_data_tuple
+        .1
+        .as_ref()
+        .expect("outputs should contain toCKB cell");
+
+    let singer_collateral = verify_capacity()?;
+    verify_data(input_data, output_data)?;
+    verify_undercollateral(singer_collateral as u128, input_data)?;
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_undercollateral/mod.rs:26]: asset_collateral = signer_collateral - xt_cell_capacity

File: contracts/toCKB-typescript/src/utils/config.rs [hunk 7]
```diff
@@ -14,3 +14,7 @@ pub const SINCE_TYPE_TIMESTAMP: u64 = 0x4000_0000_0000_0000;
 
 // 24 * 3600 means 1 day, the unit is second
 pub const SINCE_SIGNER_TIMEOUT: u64 = LOCK_TYPE_FLAG | SINCE_TYPE_TIMESTAMP | 24 * 3600;
+
+// LIQUIDATION_COLLATERAL_PERCENT means min liquidation threshold of collateral/lot_amount
+pub const LIQUIDATION_COLLATERAL_PERCENT: u8 = 115;
+pub const CKB_DECIMAL: u128 = 10000_0000;
```
Comment (huwenchao) [inline contracts/toCKB-typescript/src/utils/config.rs:20]:
```suggestion
pub const CKB_DECIMAL: u128 = 100_000_000;
```

File: tests/src/toCKB_typescript/tests/liquidation_undercollateral.rs [hunk 8]
```diff
@@ -1 +1,192 @@
...
+        .out_point(toCKB_typescript_out_point)
+        .build();
+    let always_success_lockscript = context
+        .build_script(&always_success_out_point, Default::default())
+        .expect("script");
+    let always_success_lockscript_dep = CellDep::new_builder()
+        .out_point(always_success_out_point)
+        .build();
+
+    let capacity = 375_0000u64 * CKB_DECIMAL + XT_CELL_CAPACITY;
```
Comment (huwenchao) [inline tests/src/toCKB_typescript/tests/liquidation_undercollateral.rs:150]:
```suggestion
    let capacity = 3_750_000u64 * CKB_DECIMAL + XT_CELL_CAPACITY;
```

File: tests/src/toCKB_typescript/utils/config.rs [hunk 9]
```diff
@@ -1,8 +1,20 @@
...
 // since
 pub const LOCK_TYPE_FLAG: u64 = 1 << 63;
 pub const SINCE_TYPE_TIMESTAMP: u64 = 0x4000_0000_0000_0000;
 
 // 24 * 3600 means 1 day, the unit is second
 pub const SINCE_SIGNER_TIMEOUT: u64 = LOCK_TYPE_FLAG | SINCE_TYPE_TIMESTAMP | 24 * 3600;
+
+// LIQUIDATION_COLLATERAL_PERCENT means min liquidation threshold of collateral/lot_amount
+pub const LIQUIDATION_COLLATERAL_PERCENT: u8 = 115;
+pub const CKB_DECIMAL: u64 = 10000_0000;
```
Comment (huwenchao) [inline tests/src/toCKB_typescript/utils/config.rs:20]:
```suggestion
pub const CKB_DECIMAL: u64 = 100_000_000;
```

**Review by huwenchao — APPROVED** [review]

**Review by mkxbl — APPROVED** [review]

**Pull request merged**